appsec/src/helper/service.hpp

```diff
@@ -21,13 +24,109 @@ namespace dds {
 using namespace std::chrono_literals;
 
 class service {
-public:
+protected:
+    class MetricsImpl : public metrics::TelemetrySubmitter {
+        struct tel_metric {
+            tel_metric(std::string_view name, double value, std::string tags)
+                : name{name}, value{value}, tags{std::move(tags)}
+            {}
+            std::string_view name;
+            double value;
+            std::string tags;
+        };
+
+    public:
+        MetricsImpl() = default;
+        MetricsImpl(const MetricsImpl &) = delete;
+        MetricsImpl &operator=(const MetricsImpl &) = delete;
+        MetricsImpl(MetricsImpl &&) = delete;
+        MetricsImpl &operator=(MetricsImpl &&) = delete;
+
+        ~MetricsImpl() override = default;
+
+        void submit_metric(std::string_view metric_name, double value,
+            std::string tags) override
+        {
+            SPDLOG_TRACE("submit_metric: {} {} {}", metric_name, value, tags);
+            const std::lock_guard<std::mutex> lock{pending_metrics_mutex_};
+            pending_metrics_.emplace_back(metric_name, value, std::move(tags));
+        }
+
+        void submit_legacy_metric(std::string_view name, double value) override
+        {
+            SPDLOG_TRACE("submit_legacy_metric: {} {}", name, value);
+            const std::lock_guard<std::mutex> lock{legacy_metrics_mutex_};
+            legacy_metrics_[name] = value;
+        }
+        void submit_legacy_meta(
+            std::string_view name, std::string value) override
+        {
+            SPDLOG_TRACE("submit_legacy_meta: {} {}", name, value);
+            const std::lock_guard<std::mutex> lock{meta_mutex_};
+            meta_[std::string{name}] = std::move(value);
+        }
+
+        // NOLINTNEXTLINE(bugprone-easily-swappable-parameters)
+        void submit_legacy_meta_copy_key(
+            std::string name, std::string value) override
+        {
+            const std::lock_guard<std::mutex> lock{meta_mutex_};
+            meta_[std::move(name)] = std::move(value);
+        }
+
+        template <typename Func> void drain_metrics(Func &&func)
+        {
+            std::vector<tel_metric> metrics;
+            {
+                const std::lock_guard<std::mutex> lock(pending_metrics_mutex_);
+                metrics.swap(pending_metrics_);
+            }
+            for (auto &metric : metrics) {
```

Comment:
```suggestion
            for (auto &[name, value, tags] : metrics) {
```

Reply:
Structured bindings are more readable though and are less sensitive to change, compared to... for example a hypothetical string of tags with an undefined format, as the compiler will easily let you know.

Reply:
besides what i said already, it may boil down to personal preference, but for me

```
for (auto &metric : metrics) {
  std::invoke(std::forward<Func>(func), metric.name, metric.value,
    std::move(metric.tags));
 }
 ```

reads better than

```
for (auto &[name, value, tags] : metrics) {
  std::invoke(std::forward<Func>(func), name, value, std::move(tags));
 }
 ```

because it's immediately obvious that name, value and tags refer to the metric.

Reply:
That's okay then, it's not a big deal.